File: packages/vela-sdk-ts/src/mcp/response-builder.ts
/**
 * Response building helpers for workflow tool outputs.
 *
 * Constructs structured response dicts that MCP tools return to the LLM,
 * including next-action instructions based on step type.
 */

import type { AdvanceResult, WorkflowRunState } from "../engine/types.js";
import type { IWorkflowEngine } from "../engine/workflow-engine.js";
import type { ResourceResolver } from "../engine/prompt-builder.js";
import type { Locale } from "../locale/locale.js";
import { getLocale } from "../locale/locale.js";
import type { AnyStepDefinition, WorkflowDefinition } from "../schemas/workflow.js";
import type { WorkflowStore } from "../storage/store.js";
import type { WorkflowResolver } from "./protocols.js";

// ---------------------------------------------------------------------------
// toJson
// ---------------------------------------------------------------------------

export function toJson(obj: unknown): string {
  if (typeof obj === "object" && obj !== null) {
    return JSON.stringify(obj);
  }
  return String(obj);
}

// ---------------------------------------------------------------------------
// runToDict
// ---------------------------------------------------------------------------

export function runToDict(run: WorkflowRunState): Record<string, unknown> {
  return {
    run_id: run.id,
    workflow_id: run.workflowId,
    workflow_version: run.workflowVersion,
    current_step: run.currentStep,
    status: run.status ?? null,
    project_id: run.projectId ?? null,
    params: run.params,
    state_data: run.stateData,
    parent_run_id: run.parentRunId ?? null,
    started_at: run.startedAt ?? null,
    updated_at: run.updatedAt ?? null,
  };
}

// ---------------------------------------------------------------------------
// buildNextAction
// ---------------------------------------------------------------------------

export function buildNextAction(
  run: WorkflowRunState,
  wfDef: WorkflowDefinition,
  engine: IWorkflowEngine,
  toolName: string,
  locale?: Locale,
): string {
  const loc = locale ?? getLocale();

  if (run.status === "completed") {
    return loc.workflowCompleted;
  }

  const stepDef = engine.getStep(wfDef, run.currentStep ?? "");
  if (!stepDef) {
    return loc.workflowCompleted;
  }

  const state = run.stateData;

  const captured: Record<string, unknown> = {};
  const missingKeys: string[] = [];
  if (stepDef.capture) {
    for (const cap of stepDef.capture) {
      if (cap.key in state) {
        captured[cap.key] = state[cap.key];
      } else {
        missingKeys.push(cap.key);
      }
    }
  }

  let partialHint = "";
  if (Object.keys(captured).length > 0 && missingKeys.length > 0) {
    const capturedJson = JSON.stringify(captured);
    partialHint = loc.alreadyCaptured
      .replace("{captured_json}", capturedJson)
      .replace("{missing_keys}", missingKeys.join(", "));
  }

  const hasElicitable = stepDef.capture
    ? stepDef.capture.some((c) => c.elicit !== "never")
    : false;
  const automation =
    run.params["automation_mode"] === true ||
    run.params["automation_mode"] === "true" ||
    run.params["automation_mode"] === 1;

  const fmt = (template: string): string =>
    template
      .replace(/\{tool_name\}/g, toolName)
      .replace(/\{run_id\}/g, run.id)
      .replace(/\{partial_hint\}/g, partialHint);

  if (stepDef.type === "execute") {
    // Check if this step delegates to a subagent
    if ("delegate" in stepDef && stepDef.delegate) {
      return fmt(loc.executeDelegateSubagent);
    }
    const prefixTag = automation ? loc.executePrefixTag : "";
    return fmt(loc.executeTaskThenCall).replace("{prefix_tag}", prefixTag);
  }

  if (stepDef.type === "dialog") {
    const dialogPhase = state["_dialog_phase"];
    if (automation) {
      return fmt(dialogPhase ? loc.dialogAutoProcess : loc.dialogAutoStart);
    }
    return fmt(dialogPhase ? loc.dialogConverse : loc.dialogStart);
  }

  if (hasElicitable) {
    return fmt(automation ? loc.elicitAuto : loc.elicitManual);
  }

  if (stepDef.type === "choice" && "options" in stepDef && stepDef.options.length > 0) {
    const optionsStr = stepDef.options
      .map((o) => `"${o.key}"`)
      .join(", ");
    const template = automation ? loc.choiceAuto : loc.choiceManual;
    return fmt(template).replace("{options_str}", optionsStr);
  }

  if (stepDef.type === "confirm") {
    return fmt(automation ? loc.confirmAuto : loc.confirmManual);
  }

  if (stepDef.type === "workflow") {
    const wfRef =
      ("workflow_ref" in stepDef ? stepDef.workflow_ref : null) ??
      "sub-workflow";
    return fmt(loc.subWorkflowStart).replace("{wf_ref}", wfRef);
  }

  return fmt(automation ? loc.fallbackAuto : loc.fallbackManual);
}

// ---------------------------------------------------------------------------
// buildResponse
// ---------------------------------------------------------------------------

export function buildResponse(
  result: AdvanceResult,
  wfDef: WorkflowDefinition,
  engine: IWorkflowEngine,
  toolName: string,
  locale?: Locale,
): Record<string, unknown> {
  const d: Record<string, unknown> = {
    run_id: result.run.id,
    current_step: result.run.currentStep,
    status: result.run.status ?? null,
    completed: result.completed,
  };

  if (result.prompt) {
    d["prompt"] = result.prompt;
  }

  if (result.subWorkflowRef) {
    d["sub_workflow"] = {
      ref: result.subWorkflowRef,
      params: result.subWorkflowParams,
    };
  }

  if (result.delegate) {
    d["delegate"] = result.delegate;
    if (result.delegateInstructions) {
      d["delegate_instructions"] = result.delegateInstructions;
    }
    if (result.delegateTools) {
      d["delegate_tools"] = result.delegateTools;
    }
  }

  d["next_action"] = buildNextAction(
    result.run,
    wfDef,
    engine,
    toolName,
    locale,
  );

  return d;
}

// ---------------------------------------------------------------------------
// buildStepResponse
// ---------------------------------------------------------------------------

export function buildStepResponse(
  run: WorkflowRunState,
  wfDef: WorkflowDefinition,
  engine: IWorkflowEngine,
  resolver: ResourceResolver | undefined,
  toolName: string,
  status: string = "active",
  locale?: Locale,
): Record<string, unknown> {
  const prompt = engine.assemblePrompt(wfDef, run, undefined, resolver);
  return {
    status,
    run_id: run.id,
    workflow_id: wfDef.id,
    current_step: run.currentStep,
    prompt,
    next_action: buildNextAction(run, wfDef, engine, toolName, locale),
  };
}

// ---------------------------------------------------------------------------
// enrichSubWorkflowResponse
// ---------------------------------------------------------------------------

export async function enrichSubWorkflowResponse(
  response: Record<string, unknown>,
  result: AdvanceResult,
  resolver: WorkflowResolver | undefined | null,
  store: WorkflowStore | undefined | null,
  locale: Locale | undefined,
  toolName: string,
): Promise<void> {
  if (!result.subWorkflowRef || !resolver) {
    return;
  }

  const childWfDef = await resolver.getWorkflow(result.subWorkflowRef);
  if (!childWfDef) {
    return;
  }

  const loc = locale ?? getLocale();

  let subWfInfo = (response["sub_workflow"] as Record<string, unknown>) ?? {
    ref: result.subWorkflowRef,
    params: result.subWorkflowParams,
  };

  const parentData: Record<string, unknown> = {
    ...result.run.params,
    ...result.run.stateData,
  };
  const paramsSchema: Record<string, unknown>[] = [];
  let hasIdentity = false;

  for (const p of childWfDef.params) {
    const pInfo: Record<string, unknown> = {
      name: p.name,
      required: p.required,
    };
    if (p.label) pInfo["label"] = p.label;
    if (p.description) pInfo["description"] = p.description;
    if (p.default !== undefined && p.default !== null) {
      pInfo["default"] = p.default;
    }
    const mapping = (result.subWorkflowParams ?? {}) as Record<string, string>;
    const parentKey = mapping[p.name] ?? p.name;
    if (parentKey in parentData) {
      pInfo["resolved_value"] = parentData[parentKey];
    }
    pInfo["identity"] = p.identity;
    if (p.identity) hasIdentity = true;
    paramsSchema.push(pInfo);
  }

  subWfInfo["params"] = paramsSchema;

  // Include active runs when child has identity params
  if (hasIdentity && store) {
    try {
      const activeRuns = await store.listActive({
        workflowId: childWfDef.id,
      });
      if (activeRuns.length > 0) {
        subWfInfo["active_runs"] = activeRuns.map((r) => ({
          run_id: r.id,
          status: r.status ?? null,
          current_step: r.currentStep,
          params: r.params,
        }));
      }
    } catch {
      // Ignore lookup failures
    }
  }

  response["sub_workflow"] = subWfInfo;

  // Build enriched next_action with param info for WORKFLOW steps
  if (paramsSchema.length > 0) {
    const wfRef = result.subWorkflowRef;
    const paramLines: string[] = [];
    for (const p of paramsSchema) {
      const parts: string[] = [`  - \`${p["name"]}\``];
      if (p["label"]) parts.push(`(${p["label"]})`);
      if (p["required"]) parts.push("[required]");
      if (p["identity"]) parts.push("[identity]");
      if ("resolved_value" in p) {
        parts.push(`= "${p["resolved_value"]}"`);
      } else if (p["default"] !== undefined && p["default"] !== null) {
        parts.push(`default: "${p["default"]}"`);
      }
      if (p["description"]) parts.push(`\u2014 ${p["description"]}`);
      paramLines.push(parts.join(" "));
    }

    let activeHint = "";
    const subWfRuns = subWfInfo["active_runs"] as unknown[] | undefined;
    if (subWfRuns && subWfRuns.length > 0) {
      activeHint = loc.subWorkflowActiveRunsHint.replace(
        "{count}",
        String(subWfRuns.length),
      );
    }

    response["next_action"] = loc.subWorkflowEnrichedNextAction
      .replace("{wf_ref}", wfRef)
      .replace("{param_lines}", paramLines.join("\n"))
      .replace("{active_hint}", activeHint)
      .replace("{run_id}", result.run.id)
      .replace("{tool_name}", toolName);
  }
}

// ---------------------------------------------------------------------------
// buildRunOptions
// ---------------------------------------------------------------------------

export function buildRunOptions(
  wfDef: WorkflowDefinition,
  activeRuns: WorkflowRunState[],
  locale?: Locale,
): Record<string, Record<string, string>> {
  const loc = locale ?? getLocale();

  const stepNames: Record<string, string> = {};
  for (const s of wfDef.steps) {
    stepNames[s.id] = s.name ?? s.id;
  }

  const paramLabels: Record<string, string> = {};
  for (const p of wfDef.params) {
    paramLabels[p.name] = p.label ?? p.name;
  }

  const options: Record<string, Record<string, string>> = {};
  for (const run of activeRuns) {
    const runParams = run.params;
    const paramParts: string[] = [];
    for (const pDef of wfDef.params) {
      if (pDef.name in runParams) {
        paramParts.push(
          `${paramLabels[pDef.name]}: ${runParams[pDef.name]}`,
        );
      }
    }
    const labelDetail =
      paramParts.length > 0 ? paramParts.join(", ") : run.id.slice(0, 8);
    const stepLabel = run.currentStep
      ? stepNames[run.currentStep] ?? run.currentStep
      : "";
    const stepInfo = stepLabel ? ` \u2014 ${stepLabel}` : "";
    options[run.id] = { title: `${labelDetail}${stepInfo}` };
  }

  options["__new__"] = { title: loc.newSession };
  return options;
}

// ---------------------------------------------------------------------------
// enrichToolRequirements
// ---------------------------------------------------------------------------

export function enrichToolRequirements(
  resp: Record<string, unknown>,
  wfDef: WorkflowDefinition,
  stepDef?: AnyStepDefinition | null,
): void {
  if (wfDef.tools.length > 0) {
    resp["required_tools"] = wfDef.tools.map((t) => ({
      name: t.name,
      server: t.server ?? null,
      description: t.description ?? null,
      required: t.required,
    }));
  }
  if (stepDef && stepDef.tools.length > 0) {
    resp["step_tools"] = stepDef.tools;
  }
}

File: packages/vela-sdk/src/vela_sdk/fastmcp/response_builder.py
"""Response building helpers for workflow tool outputs."""

import json
import logging
from typing import Any, Callable, Optional

from vela_sdk.engine.types import AdvanceResult, WorkflowRunState
from vela_sdk.engine.workflow_engine import WorkflowEngine
from vela_sdk.fastmcp.locale import Locale, get_locale
from vela_sdk.fastmcp.protocols import WorkflowResolver
from vela_sdk.schemas.workflow import AnyStepDefinition, StepType, WorkflowDefinition
from vela_sdk.storage.protocol import WorkflowStore

logger = logging.getLogger(__name__)


def to_json(obj: Any) -> str:
    """Serialize object to JSON string."""
    if isinstance(obj, (dict, list)):
        return json.dumps(obj, default=str, ensure_ascii=False)
    return str(obj)


def run_to_dict(run: WorkflowRunState) -> dict:
    """Convert a WorkflowRunState to a dict."""
    return {
        "run_id": run.id,
        "workflow_id": run.workflow_id,
        "workflow_version": run.workflow_version,
        "current_step": run.current_step,
        "status": run.status.value if run.status else None,
        "project_id": run.project_id,
        "params": run.params,
        "state_data": run.state_data,
        "parent_run_id": run.parent_run_id,
        "started_at": str(run.started_at) if run.started_at else None,
        "updated_at": str(run.updated_at) if run.updated_at else None,
    }


def build_next_action(
    run: WorkflowRunState,
    wf_def: WorkflowDefinition,
    engine: WorkflowEngine,
    tool_name: str,
    locale: Optional[Locale] = None,
) -> str:
    """Build a clear next_action instruction based on step type and partial state.

    Args:
        run: Current workflow run state.
        wf_def: Workflow definition.
        engine: Workflow engine instance.
        tool_name: Full advance tool name (e.g. ``"vela_advance_workflow"``).
        locale: Locale for message templates.
    """
    if locale is None:
        locale = get_locale()

    if run.status and run.status.value == "completed":
        return locale.workflow_completed

    step_def = engine._get_step(wf_def, run.current_step)
    if not step_def:
        return locale.workflow_completed

    state = run.state_data

    captured = {}
    missing_keys = []
    if step_def.capture:
        for cap in step_def.capture:
            if cap.key in state:
                captured[cap.key] = state[cap.key]
            else:
                missing_keys.append(cap.key)

    partial_hint = ""
    if captured and missing_keys:
        captured_json = json.dumps(captured, ensure_ascii=False)
        partial_hint = locale.already_captured.format(
            captured_json=captured_json,
            missing_keys=", ".join(missing_keys),
        )

    has_elicitable = any(c.elicit != "never" for c in step_def.capture) if step_def.capture else False
    automation = run.params.get("automation_mode") in (True, "true", 1)

    fmt = dict(tool_name=tool_name, run_id=run.id, partial_hint=partial_hint)

    if step_def.type == StepType.EXECUTE:
        prefix_tag = locale.execute_prefix_tag if automation else ""
        return locale.execute_task_then_call.format(prefix_tag=prefix_tag, **fmt)
    elif step_def.type == StepType.DIALOG:
        dialog_phase = state.get("_dialog_phase")
        if automation:
            if not dialog_phase:
                return locale.dialog_auto_start.format(**fmt)
            else:
                return locale.dialog_auto_process.format(**fmt)
        else:
            if not dialog_phase:
                return locale.dialog_start.format(**fmt)
            else:
                return locale.dialog_converse.format(**fmt)
    elif has_elicitable:
        if automation:
            return locale.elicit_auto.format(**fmt)
        else:
            return locale.elicit_manual.format(**fmt)
    elif step_def.type == StepType.CHOICE and step_def.options:
        options_str = ", ".join(f"\"{o.key}\"" for o in step_def.options)
        if automation:
            return locale.choice_auto.format(options_str=options_str, **fmt)
        else:
            return locale.choice_manual.format(options_str=options_str, **fmt)
    elif step_def.type == StepType.CONFIRM:
        if automation:
            return locale.confirm_auto.format(**fmt)
        else:
            return locale.confirm_manual.format(**fmt)
    elif step_def.type == StepType.WORKFLOW:
        wf_ref = getattr(step_def, "workflow_ref", None) or "sub-workflow"
        return locale.sub_workflow_start.format(wf_ref=wf_ref, **fmt)
    else:
        if automation:
            return locale.fallback_auto.format(**fmt)
        else:
            return locale.fallback_manual.format(**fmt)


def build_response(
    result: AdvanceResult,
    wf_def: WorkflowDefinition,
    engine: WorkflowEngine,
    tool_name: str,
    locale: Optional[Locale] = None,
) -> dict:
    """Build unified response dict from an AdvanceResult."""
    d = {
        "run_id": result.run.id,
        "current_step": result.run.current_step,
        "status": result.run.status.value if result.run.status else None,
        "completed": result.completed,
    }
    if result.prompt:
        d["prompt"] = result.prompt
    if result.sub_workflow_ref:
        d["sub_workflow"] = {
            "ref": result.sub_workflow_ref,
            "params": result.sub_workflow_params,
        }
    d["next_action"] = build_next_action(
        result.run, wf_def, engine, tool_name, locale=locale,
    )
    return d


async def enrich_sub_workflow_response(
    response: dict,
    result: AdvanceResult,
    resolver: Optional[WorkflowResolver] = None,
    store: Optional[WorkflowStore] = None,
    locale: Optional[Locale] = None,
    *,
    tool_name: str,
) -> None:
    """Enrich *response* in-place with child workflow param schemas and active runs.

    When ``result.sub_workflow_ref`` is set this looks up the child workflow
    definition (via *resolver*) and adds:

    * ``sub_workflow.params`` — full parameter schema with *required*, *identity*,
      *default*, and *resolved_value* (resolved from the parent run context).
    * ``sub_workflow.active_runs`` — active child runs when the child workflow has
      identity parameters, enabling the LLM to detect whether an existing run
      should be resumed instead of starting a new one.

    The function is a no-op when ``result.sub_workflow_ref`` is not set or when
    *resolver* is ``None``.
    """
    if not result.sub_workflow_ref or resolver is None:
        return

    child_wf_def = await resolver.get_workflow(result.sub_workflow_ref)
    if child_wf_def is None:
        return

    if locale is None:
        locale = get_locale()

    sub_wf_info: dict = response.get("sub_workflow", {
        "ref": result.sub_workflow_ref,
        "params": result.sub_workflow_params,
    })

    parent_data = {**result.run.params, **result.run.state_data}
    params_schema: list[dict] = []
    has_identity = False

    for p in child_wf_def.params:
        p_info: dict = {
            "name": p.name,
            "required": p.required,
        }
        if p.label:
            p_info["label"] = p.label
        if p.description:
            p_info["description"] = p.description
        if p.default is not None:
            p_info["default"] = p.default
        mapping = result.sub_workflow_params or {}
        parent_key = mapping.get(p.name, p.name)
        if parent_key in parent_data:
            p_info["resolved_value"] = parent_data[parent_key]
        p_info["identity"] = p.identity
        if p.identity:
            has_identity = True
        params_schema.append(p_info)

    sub_wf_info["params"] = params_schema

    # Include active runs when child has identity params so the LLM can
    # detect whether an existing run should be resumed.
    if has_identity and store is not None:
        try:
            active_runs = await store.list_active(workflow_id=child_wf_def.id)
            if active_runs:
                sub_wf_info["active_runs"] = [
                    {
                        "run_id": r.id,
                        "status": r.status.value if r.status else None,
                        "current_step": r.current_step,
                        "params": r.params,
                    }
                    for r in active_runs
                ]
        except Exception as e:
            logger.debug("sub_workflow.active_runs lookup failed: %s", e)

    response["sub_workflow"] = sub_wf_info

    # Build enriched next_action with param info for WORKFLOW steps
    if params_schema:
        wf_ref = result.sub_workflow_ref
        param_lines = []
        for p in params_schema:
            parts = [f"  - `{p['name']}`"]
            if p.get("label"):
                parts.append(f"({p['label']})")
            if p.get("required"):
                parts.append("[required]")
            if p.get("identity"):
                parts.append("[identity]")
            if "resolved_value" in p:
                parts.append(f"= \"{p['resolved_value']}\"")
            elif p.get("default") is not None:
                parts.append(f"default: \"{p['default']}\"")
            if p.get("description"):
                parts.append(f"— {p['description']}")
            param_lines.append(" ".join(parts))

        active_hint = ""
        if sub_wf_info.get("active_runs"):
            runs = sub_wf_info["active_runs"]
            active_hint = locale.sub_workflow_active_runs_hint.format(count=len(runs))

        response["next_action"] = (
            locale.sub_workflow_enriched_next_action.format(
                wf_ref=wf_ref,
                param_lines="\n".join(param_lines),
                active_hint=active_hint,
                run_id=result.run.id,
                tool_name=tool_name,
            )
        )


def build_step_response(
    run: WorkflowRunState,
    wf_def: WorkflowDefinition,
    engine: WorkflowEngine,
    resolver: Optional[Callable],
    tool_name: str,
    status: str = "active",
    locale: Optional[Locale] = None,
) -> dict:
    """Build response dict for current step."""
    prompt = engine.assemble_prompt(wf_def, run, resource_resolver=resolver)
    return {
        "status": status,
        "run_id": run.id,
        "workflow_id": wf_def.id,
        "current_step": run.current_step,
        "prompt": prompt,
        "next_action": build_next_action(
            run, wf_def, engine, tool_name, locale=locale,
        ),
    }


def build_run_options(
    wf_def: WorkflowDefinition,
    active_runs: list[WorkflowRunState],
    locale: Optional[Locale] = None,
) -> dict[str, dict[str, str]]:
    """Build elicit options dict from active runs + 'new' entry."""
    if locale is None:
        locale = get_locale()

    step_names = {s.id: s.name or s.id for s in wf_def.steps}
    param_labels = {p.name: p.label or p.name for p in wf_def.params}

    options: dict[str, dict[str, str]] = {}
    for run in active_runs:
        run_params = run.params
        param_parts = []
        for p_def in wf_def.params:
            if p_def.name in run_params:
                param_parts.append(f"{param_labels[p_def.name]}: {run_params[p_def.name]}")
        label_detail = ", ".join(param_parts) if param_parts else run.id[:8]
        step_label = step_names.get(run.current_step, run.current_step) if run.current_step else ""
        step_info = f" — {step_label}" if step_label else ""
        options[run.id] = {"title": f"{label_detail}{step_info}"}
    options["__new__"] = {"title": locale.new_session}
    return options


def enrich_tool_requirements(
    resp: dict,
    wf_def: WorkflowDefinition,
    step_def: Optional[AnyStepDefinition] = None,
) -> None:
    """Enrich *resp* in-place with workflow-level and step-level tool info."""
    if wf_def.tools:
        resp["required_tools"] = [
            {
                "name": t.name,
                "server": t.server,
                "description": t.description,
                "required": t.required,
            }
            for t in wf_def.tools
        ]
    if step_def and step_def.tools:
        resp["step_tools"] = step_def.tools
